poly/src/polynomial/mod.rs

```diff
@@ -70,11 +81,24 @@ pub trait MVPolynomial<F: Field>: Polynomial<F> {
 
     /// Returns the terms of a `self` as a list of tuples of the form `(Self::Term, coeff)`
     fn terms(&self) -> &[(F, Self::Term)];
+}
 
-    /// Returns the number of variables in `self`
-    fn num_vars(&self) -> usize;
+/// Describes interface for multilinear polynomials in evaluation form.
+///
+/// Index represents a point, which is a vector in {0,1}^`num_vars` in little endian form. For example, `0b1011` represents `P(1,1,0,1)`
+pub trait MultilinearPolynomialEvaluationForm<F: Field>:
+    MVPolynomial<F> + Index<usize> + IntoIterator<Item = (usize, F)>
+{
+    /// The type of partial evaluation point vectors for this polynomial.
+    type PartialPoint: Sized + Clone + Ord + Debug + Sync + Hash;
```

Comment:
Why are these called partial points / partial evaluations? Is the idea that you take a multi-linear polynomial, fix one variable, and then get a `MultilinearPolynomialEvaluationForm` object?

Reply:
Yes. Given a `PartialPoint` of dimension n, `partial_evaluate` will fix the first `n` variables and return a `MultilinearPolynomialEvaluationForm` object. Is it better to call this method `fix_variables` instead?

Reply:
Yeah, I think `fix_variables` is a better name. I'm confused as to why this is a separate struct though. E.g. why not make the API `poly.fix_variables({variable assignment vector}) -> MultilinearPolynomial`

Reply:
I named it `MultilinearPolynomialEvaluationForm` because as the name suggests, the polynomial is represented by its evaluation over {0,1}^num_vars. Many optimizations (in sumcheck, multilinear polynomial commitment scheme, etc) on multilinear polynomial only works on the evaluation form. If I name it as `MultilinearPolynomial`, it may also include multilinear polynomial in coefficient form. If a multilinear polynomial struct is in coefficient form, then this polynomial can implement `MVPolynomialCoefficientForm` instead of `MultilinearPolynomialEvaluationForm`. 

Not sure though, but does `Multilinear Polynomial` usually means a multilinear extension of an array (i.e. the evaluation form)? If so, then I can name it as `MultilinearPolynomial` instead.

Reply:
Also, I believe we can put `fix_variables` (Reduce the number of variables of the `self` by fixing the `partial_point.len()` variables at `partial_point`.) in `MVPolynomial` instead of `MultilinearPolynomialEvaluationForm` in future PR. Any idea? @ryanleh

Reply:
I see, I misunderstood what Evaluation form meant here. I don't understand why this is a trait not a struct though. How does a `MultilinearPolynomialEvaluationForm` trait differ from a `MultilinearPolynomialCoefficientForm` trait?

It'd be nice if we can have a consistent methodology for going between evaluations and coefficient representations for the univariate polynomials and multilinear polynomials.

Reply:
Like these concepts of partial evaluation, and relabeling seem like they apply for all multivariate polynomial in coefficient form

Reply:
Like these idea of partial evaluation, and relabeling seem like they apply for all multivariate polynomial in coefficient form

Reply:
I made `MultilinearPolynomialEvaluationForm` a trait because there are two structs `DenseMultilinearPolynomial` and `SparseMultilinearPolynomial` that implements this trait. I shall probably rename it as `DenseMultilinearExtension` and `SparseMultilinearExtension`. 

If a multilinear polynomial struct is in coefficient form, I believe it should implement `MVPolynomialCoefficientForm` (if multilinear coefficient form has any special optimization, we can add it in future PR). 

For now, [univariate polynomial evaluation](https://github.com/arkworks-rs/algebra/blob/3856133923c148418ef21bc0eb456ba69185ca5f/poly/src/evaluations.rs#L18) form does not implement `Polynomial` trait. I think for this PR I should not let `MultilinearPolynomialEvaluationForm` extends `Polynomial` trait either. We can unify those APIs in later PR. What do you think?

Reply:
I think I'm now understanding what you mean. This PR is only implementing operations on the evaluations of a multi-linear polynomial. 

I'm still confused about this as the trait though. Don't the evaluations need an associated domain, to specify what points in the field the polynomial is being evaluated over? This is part of why If that was included, then I get having the evaluation form API. Also shouldn't the `relabel` and `fix variables` methods be pushed to `MVPolynomial`? (We can leave the implementation as stubs/panics in `MVSparsePolynomial` for now)

Also agreed re the univariate polynomial evaluation form. Maybe we should rethink it to be more along the lines of what you're doing in this PR.

Reply:
I've done most of the refactoring, but I will check it again tomorrow. Once I finish, I will reopen this PR